File: backend/normalizer.py
import json
import unicodedata
from pathlib import Path
from functools import lru_cache

SINONIMOS_PATH = Path(__file__).parent / "data" / "sinonimos.json"


def quitar_acentos(texto: str) -> str:
    nfkd = unicodedata.normalize("NFKD", texto)
    return "".join(c for c in nfkd if not unicodedata.combining(c))


def limpiar(texto: str) -> str:
    return quitar_acentos(texto.strip().lower())


@lru_cache(maxsize=1)
def cargar_mapa_sinonimos() -> dict[str, str]:
    with open(SINONIMOS_PATH, encoding="utf-8") as f:
        sinonimos = json.load(f)

    mapa = {}
    for canonico, alias_list in sinonimos.items():
        clave_canonica = limpiar(canonico)
        mapa[clave_canonica] = clave_canonica
        for alias in alias_list:
            mapa[limpiar(alias)] = clave_canonica
    return mapa


def normalizar_ingrediente(texto: str) -> str:
    if not texto:
        return ""
    limpio = limpiar(texto)
    mapa = cargar_mapa_sinonimos()
    return mapa.get(limpio, limpio)


def normalizar_lista(textos: list[str]) -> list[str]:
    vistos = set()
    resultado = []
    for t in textos:
        norm = normalizar_ingrediente(t)
        if norm and norm not in vistos:
            vistos.add(norm)
            resultado.append(norm)
    return resultado


def parsear_input_usuario(texto: str) -> list[str]:
    separadores = [",", ";", "\n", "/", "|", " y ", " con ", " mas ", "+", "&"]
    partes = [texto]
    for sep in separadores:
        nuevas = []
        for p in partes:
            nuevas.extend(p.split(sep))
        partes = nuevas
    return normalizar_lista([p for p in partes if p.strip()])

File: frontend/src/lib/recipes/normalizer.js
import sinonimos from './sinonimos.json';

function quitarAcentos(texto) {
  return texto.normalize('NFD').replace(/[̀-ͯ]/g, '');
}

function limpiar(texto) {
  return quitarAcentos(String(texto || '').trim().toLowerCase());
}

let mapaCache = null;

function cargarMapa() {
  if (mapaCache) return mapaCache;
  const mapa = {};
  for (const [canonico, aliases] of Object.entries(sinonimos)) {
    const claveCanonica = limpiar(canonico);
    mapa[claveCanonica] = claveCanonica;
    for (const alias of aliases) {
      mapa[limpiar(alias)] = claveCanonica;
    }
  }
  mapaCache = mapa;
  return mapa;
}

export function normalizarIngrediente(texto) {
  if (!texto) return '';
  const limpio = limpiar(texto);
  return cargarMapa()[limpio] || limpio;
}

export function normalizarLista(textos) {
  const vistos = new Set();
  const resultado = [];
  for (const t of textos || []) {
    const norm = normalizarIngrediente(t);
    if (norm && !vistos.has(norm)) {
      vistos.add(norm);
      resultado.push(norm);
    }
  }
  return resultado;
}

export function parsearInputUsuario(texto) {
  if (!texto) return [];
  const separadores = [',', ';', '\n', '/', '|', ' y ', ' con ', ' mas ', '+', '&'];
  let partes = [texto];
  for (const sep of separadores) {
    const nuevas = [];
    for (const p of partes) nuevas.push(...p.split(sep));
    partes = nuevas;
  }
  return normalizarLista(partes.filter((p) => p.trim()));
}
